### PR TITLE
Update to Cadence v0.12.3, SDK v0.14.0, and Emulator v0.14.0-beta2

### PR DESCRIPTION
**Pull request by turbolent**
Work towards onflow/flow#309

Update the contracts, transactions, and scripts to the latest version of Cadence.

Along the way I also made some parts a bit more readable my adding line breaks, and I also refactored some parts slightly to again, make them more readable. We can revert these changes if you don't like them.

While updating the code I've written up these instructions to help explain what was migrated here and assist others with their migration. I plan to add to these instructions to the Cadence repo after getting feedback here. Maybe this could be improved? Maybe there are more examples required or the ones I have are not good enough?

---

- The return types of `PublicAccount.getCapability` and `AuthAccount.getCapability` are not optional anymore.

  For example, in the following code the force unwrapping should be removed:

  ```diff
       let balanceRef = account
  -        .getCapability(/public/flowTokenBalance)!
  +        .getCapability(/public/flowTokenBalance)
           .borrow<&FlowToken.Vault{FungibleToken.Balance}>()!
  ```

  In the next example, optional binding was used and is not allowed anymore:

  ```diff
  -    if let balanceCap = account.getCapability(/public/flowTokenBalance) {
  -        return balanceCap.borrow<&FlowToken.Vault{FungibleToken.Balance}>()!
  -    }

  +    let balanceCap = account.getCapability(/public/flowTokenBalance)
  +    return balanceCap.borrow<&FlowToken.Vault{FungibleToken.Balance}>()!
  ```

- Parameters of the Storage API functions that had the type `Path` now have more specific types.
  For example, the `getCapability` functions now require a `CapabilityPath` instead of just a `Path`.

  Ensure path values with the correct path type are passed to these functions.

  For example, a contract may have declared a field with the type `Path`, then used it in a function to call `getCapability`.
  The type of the field must be changed to the more specific type:

    ```diff
     pub contract SomeContract {

    -    pub let somethingPath: Path
    +    pub let somethingPath: StoragePath

         init() {
             self.somethingPath = /storage/something
         }

         pub fun borrow(): &Something {
             return self.account.borrow<&Something>(self.somethingPath)
         }
     }
    ```

---

The tests are not fully passing yet. From what I can see from the failures, it looks like all failing tests are due to unexpected fixed-point values, and the expected and actual values are fairly close.

Could it be that this is due to the fix for fixed-point number multiplication / division? 

Maybe the [temporary workaround](https://github.com/onflow/flow-core-contracts/blob/301206b27090fa9116c1aba35cd8bade8a26857c/contracts/FlowIDTableStaking.cdc#L716-L722) can now be removed?

Help with resolving these outstanding issues would be much appreciated. Thank you!

```
=== RUN   TestIDTable/Should_pay_correct_rewards,_no_delegators_are_paid_because_none_are_staked_yet
    TestIDTable/Should_pay_correct_rewards,_no_delegators_are_paid_because_none_are_staked_yet: flow_idtable_staking_test.go:2099:
        	Error Trace:	flow_idtable_staking_test.go:2099
        	Error:      	Not equal:
        	            	expected: 0x110d9316ec00
        	            	actual  : 0x1139abd633f0
        	Test:       	TestIDTable/Should_pay_correct_rewards,_no_delegators_are_paid_because_none_are_staked_yet
    TestIDTable/Should_pay_correct_rewards,_no_delegators_are_paid_because_none_are_staked_yet: flow_idtable_staking_test.go:2155:
        	Error Trace:	flow_idtable_staking_test.go:2155
        	Error:      	Not equal:
        	            	expected: 0x5f7f37b39000
        	            	actual  : 0x607628af8940
        	Test:       	TestIDTable/Should_pay_correct_rewards,_no_delegators_are_paid_because_none_are_staked_yet
=== RUN   TestIDTable/Should_Move_tokens_between_buckets
    TestIDTable/Should_Move_tokens_between_buckets: flow_idtable_staking_test.go:2239:
        	Error Trace:	flow_idtable_staking_test.go:2239
        	Error:      	Not equal:
        	            	expected: 0x110d9316ec00
        	            	actual  : 0x1139abd633f0
        	Test:       	TestIDTable/Should_Move_tokens_between_buckets
    TestIDTable/Should_Move_tokens_between_buckets: flow_idtable_staking_test.go:2389:
        	Error Trace:	flow_idtable_staking_test.go:2389
        	Error:      	Not equal:
        	            	expected: 0x5f7f37b39000
        	            	actual  : 0x607628af8940
        	Test:       	TestIDTable/Should_Move_tokens_between_buckets
=== RUN   TestIDTable/Should_create_new_execution_node
=== RUN   TestIDTable/Should_be_able_to_delegate_new_tokens_to_max
=== RUN   TestIDTable/Should_not_be_able_request_unstaking_below_the_minimum_if_a_node_has_delegators_staked_or_committed
=== RUN   TestIDTable/Should_be_able_to_commit_rewarded_tokens
    TestIDTable/Should_be_able_to_commit_rewarded_tokens: flow_idtable_staking_test.go:2593:
        	Error Trace:	flow_idtable_staking_test.go:2593
        	Error:      	Not equal:
        	            	expected: 0xc816bdd9c00
        	            	actual  : 0xcad849ce3f0
        	Test:       	TestIDTable/Should_be_able_to_commit_rewarded_tokens
=== RUN   TestIDTable/Should_be_able_to_request_unstake_delegated_tokens_from_Josh,_marks_as_requested
=== RUN   TestIDTable/Should_be_able_cancel_unstake_request_for_delegator
=== RUN   TestIDTable/Should_be_able_to_request_unstake_delegated_tokens_from_Josh,_marks_as_requested#01
=== RUN   TestIDTable/Should_be_able_to_end_the_staking_auction
=== RUN   TestIDTable/Should_Move_tokens_between_buckets#01
=== RUN   TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators
    TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators: flow_idtable_staking_test.go:2868:
        	Error Trace:	flow_idtable_staking_test.go:2868
        	Error:      	Not equal:
        	            	expected: 0xc816bdd9c00
        	            	actual  : 0xcad849ce3f0
        	Test:       	TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators
    TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators: flow_idtable_staking_test.go:2884:
        	Error Trace:	flow_idtable_staking_test.go:2884
        	Error:      	Not equal:
        	            	expected: 0x5f7f37b39000
        	            	actual  : 0x607628af8940
        	Test:       	TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators
    TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators: flow_idtable_staking_test.go:2906:
        	Error Trace:	flow_idtable_staking_test.go:2906
        	Error:      	Not equal:
        	            	expected: 0xc816bdd9c00
        	            	actual  : 0xcad849ce3f0
        	Test:       	TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators
    TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators: flow_idtable_staking_test.go:2914:
        	Error Trace:	flow_idtable_staking_test.go:2914
        	Error:      	Not equal:
        	            	expected: 0x113d126fe400
        	            	actual  : 0x11d2babef0c0
        	Test:       	TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators
    TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators: flow_idtable_staking_test.go:2922:
        	Error Trace:	flow_idtable_staking_test.go:2922
        	Error:      	Not equal:
        	            	expected: 0x1167727b800
        	            	actual  : 0x11fe8aac680
        	Test:       	TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators
    TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators: flow_idtable_staking_test.go:2930:
        	Error Trace:	flow_idtable_staking_test.go:2930
        	Error:      	Not equal:
        	            	expected: 0x89969ea96000
        	            	actual  : 0x8bfafad8d840
        	Test:       	TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators
    TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators: flow_idtable_staking_test.go:2938:
        	Error Trace:	flow_idtable_staking_test.go:2938
        	Error:      	Not equal:
        	            	expected: 0x2b829e34c00
        	            	actual  : 0x2cfc5aaf040
        	Test:       	TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators
    TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators: flow_idtable_staking_test.go:2946:
        	Error Trace:	flow_idtable_staking_test.go:2946
        	Error:      	Not equal:
        	            	expected: 0x57053c69800
        	            	actual  : 0x59f8b55e080
        	Test:       	TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators
    TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators: flow_idtable_staking_test.go:2954:
        	Error Trace:	flow_idtable_staking_test.go:2954
        	Error:      	Not equal:
        	            	expected: 0x2961cb565800
        	            	actual  : 0x2ac90de67480
        	Test:       	TestIDTable/Should_pay_correct_rewards,_rewards_are_split_up_properly_between_stakers_and_delegators
=== RUN   TestIDTable/Should_Move_tokens_between_buckets#02
=== RUN   TestIDTable/Should_be_able_to_withdraw_delegator_rewards
    TestIDTable/Should_be_able_to_withdraw_delegator_rewards: flow_idtable_staking_test.go:3028:
        	Error Trace:	flow_idtable_staking_test.go:3028
        	Error:      	Not equal:
        	            	expected: 0x28998f57c00
        	            	actual  : 0x2a134bd2040
        	Test:       	TestIDTable/Should_be_able_to_withdraw_delegator_rewards
```

### CODE REVIEW

**Comment by joshuahannan** [comment]
@turbolent I'll make a PR to this branch with my suggested changes and I'll also try to fix the tests. I'm pretty sure that the incorrect results are because of the workaround, which I'll remove.

**Comment by turbolent** [comment]
@joshuahannan sounds good, thank you!

File: contracts/FlowIDTableStaking.cdc [hunk 1]
```diff
@@ -79,15 +79,15 @@ pub contract FlowIDTableStaking {
     /// value = decimal number between 0 and 1 indicating a percentage
     access(contract) var rewardRatios: {UInt8: UFix64}
 
-    /// The percentage of rewards that every node operator takes from 
+    /// The percentage of rewards that every node operator takes from
     /// the users that are delegating to it
     access(contract) var nodeDelegatingRewardCut: UFix64
 
     /// Paths for storing staking resources
-    pub let NodeStakerStoragePath: Path
```
Comment (joshuahannan) [inline contracts/FlowIDTableStaking.cdc:87]: @turbolent Is changing the type of the path here something that is going to be compatible with the way that contracts are upgraded? (cannot add or remove state)
Should I make a test to see if doing an upgrade like this works properly?

File: contracts/LockedTokens.cdc [hunk 2]
```diff
@@ -43,57 +43,57 @@ pub contract LockedTokens {
 
     pub event LockedTokensDeposited(address: Address, amount: UFix64)
 
-    /// Path to store the locked token manager resource 
+    /// Path to store the locked token manager resource
     /// in the shared account
-    pub let LockedTokenManagerStoragePath: Path
```
Comment (joshuahannan) [inline contracts/LockedTokens.cdc:48]: Same comment about changing the type of a field via a contract upgrade.

File: contracts/StakingProxy.cdc [hunk 3]
```diff
@@ -3,15 +3,15 @@
...
 // use to store staking proxies for all of their node operation
-// relationships 
+// relationships
 
 pub contract StakingProxy {
 
     /// path to store the node operator resource
     /// in the node operators account for staking helper
-    pub let NodeOperatorCapabilityStoragePath: Path
+    pub let NodeOperatorCapabilityStoragePath: StoragePath
```
Comment (joshuahannan) [inline contracts/StakingProxy.cdc:12]: Same comment about upgrading

File: contracts/FlowIDTableStaking.cdc (hunk 1 above)
Comment (turbolent) [reply contracts/FlowIDTableStaking.cdc:87]: Yes, this is just a change of the type of an existing field, which is supported. This doesn't add or remove any state.

Comment (turbolent) [reply contracts/FlowIDTableStaking.cdc:87]: Of course, the value stored in the existing field needs to be compatible with the new, more specific type, which I checked for each field.

**Comment by joshuahannan** [comment]
I just want to flag a couple things for @psiemens @laynelafrance @JeffreyDoyle. 

This PR changes transactions and scripts that are used by Flow Port and many other applications (Kraken, Blocto, Finoa, Coinlist, flowscan) for their locked token and staking processes and user interfaces. As far as I know, most of these apps have copied and pasted the original versions of these templates into their application code. Once we do the next spork, we are going to have to have everyone who uses these update the text of all of their transactions also, because old templates will not work with the newest version of cadence and their apps will not work.

I think we should add a javascript package in this repo so that doing upgrades with breaking changes like this will be easier in the future. Just like with Go, we can just tag a new release and inform other apps that they need to upgrade to the latest version.

Thoughts?

**Comment by psiemens** [comment]
> I think we should add a javascript package in this repo so that doing upgrades with breaking changes like this will be easier in the future. Just like with Go, we can just tag a new release and inform other apps that they need to upgrade to the latest version.

Great idea, definitely on board for this!

> Once we do the next spork, we are going to have to have everyone who uses these update the text of all of their transactions

To clarify, is this set to go out in the spork on the 27th? Unfortunately this is going to require an update to all of the pre-baked Ledger transactions, which is more difficult than getting Kraken + Finoa to update their systems.

**Comment by turbolent** [comment]
@joshuahannan @psiemens The major change is that `getCapability` doesn't return an optional anymore. The common pattern in transaction was that it was force-unwrapped. Knowing this was done in a lot of places, I also made force-unwrapping of non-optionals a warning, so existing transactions should mainly keep working. 

Could you please point me to the transactions that are actually used and that might need to be updated?

**Comment by psiemens** [comment]
> Knowing this was done in a lot of places, I also made force-unwrapping of non-optionals a warning, it is not an error, so existing transactions should mainly keep working.

Oh great! I think we should try to get all the transactions updated ASAP, but it's good to know that we aren't in a crunch. Ledger updates can be slow.

@turbolent As a sample, here are all the staking transactions currently baked into the Ledger app: https://github.com/onflow/flow-core-contracts/tree/master/transactions/lockedTokens/staker

**Comment by turbolent** [comment]
@psiemens I looked through these and it seems like the only backwards-incompatible code is this: https://github.com/onflow/flow-core-contracts/blob/master/transactions/lockedTokens/staker/get_staker_info.cdc#L20-L27

This PR updates this transaction, if users will run the existing / not-updated transaction in the future, it would not be valid anymore.

**Pull request merged**